pallets/clusters/src/lib.rs

```diff
@@ -0,0 +1,520 @@
+			cluster_id: Hash128,
+			role_name: BoundedVec<u8, T::NameLimit>,
+			settings: RoleSetting,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!cluster_id.len().is_zero(), Error::<T>::InvalidInput);
+			ensure!(!role_name.len().is_zero(), Error::<T>::InvalidInput);
+			ensure!(!settings.name.len().is_zero(), Error::<T>::InvalidInput);
+			ensure!(!settings.data.len().is_zero(), Error::<T>::InvalidInput);
+
+			let role_hash = blake2_128(&[&cluster_id.as_slice(), role_name.as_slice()].concat());
+			let cluster = <Clusters<T>>::get(&cluster_id).ok_or(Error::<T>::ClusterNotFound)?;
+
+			// Check that the caller is the owner of the cluster
+			ensure!(
+				who == cluster.owner,
+				Error::<T>::NoPermission
+			);
+
+			// At creation there are no Members and no Rules assigned to a Role
+			let new_role = Role { name: role_name.into_inner(), members: vec![], rules: None };
+
+			// Check that the role does not exists already in the cluster
+			if cluster
+				.roles
+				.iter()
+				.any(|role| new_role.name.eq(&role.name))
+			{
+				return Err(Error::<T>::RoleExists.into())
+			}
+
+			// write
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster
+					.as_mut()
+					.expect("Should find the cluster");
+				cluster.roles.push(new_role);
+			});
+
+			<RoleToSettings<T>>::append(role_hash, settings);
+
+			Self::deposit_event(Event::RoleCreated { role_hash });
+			log::trace!("Role {:?} created and associated to cluster {:?}", role_hash, cluster_id);
+
+			Ok(())
+		}
+
+		/// Edit a **Role**.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn edit_role(
+			origin: OriginFor<T>,
+			role_name: BoundedVec<u8, T::NameLimit>,
+			cluster_id: Hash128,
+			new_members_list: BoundedVec<T::AccountId, T::MembersLimit>,
+			new_settings: RoleSetting,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!role_name.len().is_zero(), Error::<T>::InvalidInput);
+
+			if new_members_list.len().is_zero() &&
+				(new_settings.data.is_empty() || new_settings.name.is_empty())
+			{
+				return Err(Error::<T>::InvalidInput.into())
+			}
+
+			let cluster = <Clusters<T>>::get(&cluster_id).ok_or(Error::<T>::SystematicFailure)?;
+
+			// Check that the caller is the owner of the cluster;
+			ensure!(
+				who == cluster.owner,
+				Error::<T>::NoPermission
+			);
+
+			// Check that the role exists in the cluster and in storage
+			if !cluster
+				.roles
+				.iter()
+				.any(|role| role_name.eq(&role.name))
+			{
+				return Err(Error::<T>::RoleNotFound.into())
+			}
+
+			let role_hash = blake2_128(&[&cluster_id[..], &role_name.as_slice()].concat());
+			let role_name_vec = role_name.into_inner();
+
+			// write
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster
+					.as_mut()
+					.expect("Should find the cluster");
+				cluster.roles.iter().position(|x| x.name == role_name_vec).map(|index| {
+					let role = cluster
```

Comment:
you are already iterating O(n) here https://marketplace.fragnova.com/proto-details/0x79543750bd76e74245ee62f3d92a77ba53cf49d934b1cf7196a242edf2138932
Take the index as well there.

Reply:
I can add a storage that maps a `cluster_id` to a `Vec<Role>` so I could avoid the iteration. I had a similar structure in a [previous commit](https://github.com/fragcolor-xyz/clamor/pull/224/commits/75884cf583d3db70a38355f2c0edebbfe1fb7e5e#diff-6804b4445503eaa0a518d6cabdb56e2901be2c3b0c19ba2b7b4a5e4becdad173), but it was mapping a `cluster_id` to a `Vec<role_hash>` .
Would that work (map(`cluster_id`, `Vec<Role>`)?

Reply:
Forget adding storage.
My point is simpler.
You are doing `cluster.roles.iter()` twice no? (including role_name check) so you could have taken the index before already no?

Reply:
Not sure to understand. It's not doing `cluster.roles.iter()` twice. The first `iter()` is for taking the index of the element in the vector via `position` and then i take the vector `cluster.roles` and i access to the element i need by its index (O(1)), without doing any other loop over the vector.

Reply:
zoom out, way above you are doing exactly the same iteration.